### ISSUE TITLE
AttributeError: 'RasterModelGrid' object has no attribute '_gradient_boundary_node_links'

### ISSUE THREAD

**Issue by jadams15** [landlab/landlab#472]
When I am trying to use ``set_watershed_boundary_condition()``, for some reason, I am getting this error contained within ``_update_links_nodes_cells_to_new_BCs()``. Interesting to note I tested it on a different DEM without any issue. I get a similar error if I try to use ``set_nodata_nodes_to_closed()``. 

Since this doesn't exist, can it be removed? I couldn't find a deprecation warning or anything to replace it with an updated function.

**Comment by jadams15** [comment]
@mcflugen I tagged you because I think you were one of the primary architects of base.py. If not, sorry!

**Comment by mcflugen** [comment]
@jadams15 I am familiar with much of `base.py` but didn't have much to do with the boundary condition stuff. Since `_gradient_boundary_node_links` doesn't exist anywhere, I suppose that line could be safely removed but I may be missing something. Do things work as expected if you remove that line?

I'm surprised tests aren't failing and that it works for a different DEM. I guess that means `set_watershed_boundary_condition` isn't being tested or that it's being tested with a DEM that, for some reason, doesn't call `_update_links_nodes_cells_to_new_BCs`. Could you please write some tests that touch these lines?

**Comment by jadams15** [comment]
if I remove the line, things work fine. Not sure why... I can try to write a test or two for `_update_links_nodes_cells_to_new_BCs`. We'll see how it goes!

**Comment by mcflugen** [comment]
Thanks. If you could also figure out why you didn't get the error with a different DEM, that would also be great!

**Comment by jadams15** [comment]
It's actually really strange because that line shouldn't even be throwing an error, if I understand it correctly. In the try-except-else case, the "try" line is True. It should be ignoring the "except-else" part. I have spent the better part of this afternoon trying to understand why the error is specific to this DEM. The only difference was the need for a halo of perimeter nodes - and that didn't fix the error either. 

To get even stranger, it doesn't throw the error when I read in a DEM/set boundary status on the command line, only when I run it using Spyder. Super frustrating. 

At the very least, it will be tested!

**Comment by mcflugen** [comment]
The `except` part is only called if the `try` part raises and exception. If the `try` part doesn't raise an exception, only then is the `else` part executed. I guess with some DEMs `_fixed_grad_links_created` does not exist and so the `except` part is run, which means the `else` part isn't touched.

Thanks for looking into this!

**Comment by SiccarPoint** [comment]
`_update_links_nodes_cells_to_new_BCs` is a key method in `base.py`, I wouldn't recommend disabling it! It is the method that (should?) get called whenever any BC gets changed anywhere on a grid, and does all the testing to ensure internal BC consistency (e.g., FIXED links end with one CORE and one FIXED_GRADIENT node). I would guess @mcflugen is right: not all the element types (or their flags) get created when a DEM grid is produced, so the update method fails.

None of this is formally tested, largely because it was implemented so piecemeal. Sorry!

**Comment by mcflugen** [comment]
@SiccarPoint Not to worry. We aren't talking about disabling `_update_links_nodes_cells_to_new_BCs`, only the line in `base.py` that calls (the nonexistent ) `_gradient_boundary_node_links`. However, since @jadams15's grid does have the attribute `_fixed_grad_links_created`, I suppose a method like `_gradient_boundary_node_links` should be called to update the fixed gradient boundary conditions. Perhaps you left this line in as a reminder that we need to write such a function?

**Comment by SiccarPoint** [comment]
Uh… Probably?

**Comment by jadams15** [comment]
yes! we aren't disabling `_update_links_nodes_cells_to_new_BCs`. however, I am not sure if `_gradient_boundary_node_links` was needed after all - I suspect if `_create_fixed_gradient_boundary_node_links()` is called within `_update_links_nodes_cells_to_new_BCs`, it will achieve the same goal. I am working on that now. My python distribution has been giving me issues and I can't run the tests locally - trying to fix that problem before pushing the branch back to Git.
